text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+   */
+  public static fromOriginAccessControlAttributes(scope: Construct, id: string, attrs: OriginAccessControlAttributes): IOriginAccessControl {
+    class Import extends Resource implements IOriginAccessControl {
+      public readonly originAccessControlId = attrs.originAccessControlId;
+      public readonly originAccessControlOriginType = attrs.originAccessControlOriginType;
+    }
+    return new Import(scope, id);
+  }
+  /**
+   * Creates an origin access control for a S3 origin.
+   */
+  public static forS3(scope: Construct, id: string, props?: CommonOriginAccessControlProps): IOriginAccessControl {
+    return new OriginAccessControl(scope, id, {
+      ...props,
+      originAccessControlOriginType: OriginAccessControlOriginType.S3,
+    });
+  }
+
+  public readonly originAccessControlId: string;
+  public readonly originAccessControlOriginType: string;
+
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps) {
+    super(scope, id);
+    this.originAccessControlOriginType = props.originAccessControlOriginType;
+
+    // Check if origin access control name is 64 characters or less
+    if (props.originAccessControlName) {
+      this.validateOriginAccessControlName(props.originAccessControlName);
+    }
+
+    const resource = new CfnOriginAccessControl(this, 'Resource', {
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? Names.uniqueResourceName(this, {
+          maxLength: 64,
+        }),
+        signingBehavior: props.signing?.behavior ?? SigningBehavior.ALWAYS,
+        signingProtocol: props.signing?.protocol ?? SigningProtocol.SIGV4,
+        originAccessControlOriginType: this.originAccessControlOriginType,
+      },
+    });
+
+    this.originAccessControlId = resource.attrId;
+  }
+
+  private validateOriginAccessControlName(name: string) {
+    if (!Token.isUnresolved(name) && name.length > 64) {
+      throw new Error(`Origin access control name must be 64 characters or less, '${name}' has length ${name.length}`);
+    }
+  }
+}
+```
+
+#### Modifications to `S3BucketOrigin` class
+
+The `S3BucketOrigin` will have two methods, `withAccessIdentity()` and `withAccessControl()`, which each return a class configured with
+the corresponding method of origin access control.
+
+In the case where an imported bucket is being used for the S3 origin, calling `bucket.addToResourcePolicy()` will fail to add the policy statement. Existing
+[workarounds](https://github.com/aws/aws-cdk/issues/6548#issuecomment-869091553) require the user to create a new `BucketPolicy` for the bucket and
+add the policy statements using `bucketPolicy.document.addStatements()`.
+However, this overwrites the whole bucket policy instead of appending statements to the
+existing policy which is a subpar user experience. The proposed solution to this issue is
+to use a custom resource to retrieve the existing bucket policy and append the
```

Comment:
I do not think this should be the proposed solution. Two reasons:

1. Modifying resources that are not under CFN control in a CFN deployment is a bad idea. It introduces horrible edge cases and complexity (see the recent regressions in custom resources). It used to be the only alternative to deleting and recreating, but now we have CDK migrate. We should point users there instead, because modifying resources that don't live in your CFN template or CDK code is an antipattern. 
2. This opens up the door for users contributing a custom resource that does this. We should not include this in base CDK.